src/crimson/os/seastore/seastore_types.h

```diff
@@ -1117,11 +1117,44 @@ constexpr bool is_backref_node(extent_types_t type)
 
 std::ostream &operator<<(std::ostream &out, extent_types_t t);
 
+/**
+ * reclaim_gen_t
+ *
+ * The goal is to group the similar aged extents in the same segment for better
+ * bimodel utilization distribution, and also to the same device tier. For EPM,
+ * it has the flexibility to make placement decisions by re-assigning the
+ * generation. And each non-inline generation will be statically mapped to a
+ * writer in EPM.
+ *
+ * All the fresh and dirty extents start with INIT_GENERATION upon allocation,
+ * and they will be assigned to INLINE/OOL generation by EPM before the initial
+ * writes. After that, the generation can only be increased upon rewrite.
+ *
+ * Note, although EPM can re-assign the generations according to the tiering
+ * status, it cannot decrease the generation for the correctness of space
+ * reservation. It may choose to assign a larger generation if the extent is
+ * hinted cold, or if want to evict extents to the cold tier. And it may choose
+ * to not increase the generation if want to keep the hot tier as filled as
+ * possible.
```

Comment:
Adjusting mappings or introducing new metrics might work, but I still prefer to start with a simple solution that can fulfill the essential requirements of device tiering:
1. The hot tier is preferred to be full to make the most use of the costy storage device.
2. When the hot tier is filled, it needs to migrate cold data from the hot to cold tier effectively.
3. There are hints from user to mark data as cold, although cannot be 100% accurate.
4. Correctness: disk space reservation should not introduce deadlock, it is simpler to have the constraint to not move unchanged data back to lower/hotter generation/tier. If some cold data requires frequent reads, the cache should work.

I think the proposed generation usage is able to address these requirements and be relatively simpler:
1. When the hot tier is empty, EPM can assign the target `gen(5)` extents back to `gen(4)` to retain them in the hot tier, which is identical to the existing behavior without a cold tier.
2. When the hot tier is filled, EPM is free to move the hot-tier reclaiming data `gen(2,3,4)` to the cold tier by assigning the cold tier generations `gen(>= 5)`.
3. When data from user is hinted cold, EPM is free to assign a larger generation.
4. The generation won't decrease, so the unmodified data won't flow back, and the tier-level DAG reservation and blocking should work.

> I think this may break the assumption that similar aged extents are put in the same segment and overwhelm AsyncCleaner. Am I right?

If it's really an issue, then it is simple to introduce more generations in the cold tier and assign them in a finer-grained way.

Without introducing more generations in the cold tier, another consideration is that a device-tier won't change from empty to full immediately, so the mixture of aged extents in a segment won't be common. And the policy of generational GC actually ignores the gen value and only takes the modified-time and utility into the consideration.

> BTW, I still think that adjusting the mapping from generation to writer to put data in the cold tier, which you mentioned during a discussion the other day, is a good idea:-)

I cannot yet find out an intuitive and simpler way to proper implement the flexible mapping solution. For example, if the `gen(3)` writer is redirected to hot tier and cold tier from time to time, it is possible for `gen(3)` segments to be in both tiers. And if the `gen(4)` writer is also allowed to be redirected to both tiers, the cold-tier `gen(3)` data will have the chance to be moved back to the hot tier. I assume that both the flexible mapping design and the undetermined write directions makes it harder (or more complex) to implement proper reservation and blocking strategy.

Reply:
> I think the proposed generation usage is able to address these requirements and be relatively simpler:
> 
>     1. When the hot tier is empty, EPM can assign the `target gen(5)` extents back to `gen(4)` to retain them in the hot tier, which is identical to the existing behavior without a cold tier.
> 
>     2. When the hot tier is filled, EPM is free to move the hot-tier reclaiming data `gen(2,3,4)` to the cold tier by assigning the cold tier generations `gen(>= 5)`.
> 
>     3. When data from user is hinted cold, EPM is free to assign a larger generation.
> 
>     4. The generation won't decrease, so the unmodified data won't flow back, and the tier-level DAG reservation and blocking should work.

@cyx1231st Ok, I have implemented a simple strategy that directly rewrites the Data/MD `gen(2/3)` to the cold tier for #47974, addressing https://github.com/ceph/ceph/pull/47974#issuecomment-1240854954. Right now, I'm doing some final code clean work, after which I'll repush it. After that, I'll follow this suggestion to implement the dynamic generatuin assigning. Based on my current implementation, this shouldn't be a lot of work.